File: bookstore-minimal/backend/src/main/kotlin/masha/bookstore/services/ReviewService.kt
package masha.bookstore.services

import masha.bookstore.core.BaseService
import masha.bookstore.dtos.ReviewDto
import masha.bookstore.interfaces.IReviewService
import masha.bookstore.models.Review
import masha.bookstore.models.Book
import masha.bookstore.repositories.ReviewRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service

@Service
class ReviewService(
    @Autowired override val repository: ReviewRepository
) : BaseService<Review, ReviewRepository>(repository), IReviewService {

    override fun create(book: Book, dto: ReviewDto): Review {
        val review = Review(book = book, author = dto.author, description = dto.description)
        repository.save(review)
        return review
    }

    override fun change(id: Int, dto: ReviewDto): Review {
        val review: Review = getOne(id)
        review.author = dto.author
        review.description = dto.description
        repository.save(review)
        return review
    }

    override fun deleteWithBook(book: Book) {
        repository.deleteAll(book.reviews)
    }
}

File: hotel-mern/services/review.service.ts
import { Review } from '../shared/models'
import { ReviewRepository } from '../shared/repositories'

export default class ReviewService extends ReviewRepository {
  async getByRoom(_room: string): Promise<Review[]> {
    return this.model.find({ _room }).lean()
  }

  async getOne(_id: string): Promise<Review> {
    return this.model.findById(_id).lean()
  }

  async create(
    _room: string,
    author: string,
    content: string
  ): Promise<Review> {
    const review = await new this.model({ _room, author, content })
    await review.save()
    return this.getOne(review._id)
  }

  async change(_id: string, content: string): Promise<Review> {
    await this.model.findByIdAndUpdate(_id, {
      $set: { content },
    })
    return this.getOne(_id)
  }

  async delete(_id: string): Promise<string> {
    await this.model.findByIdAndRemove(_id)
    return _id
  }

  async deleteWithRoom(roomId: string): Promise<void> {
    await this.model.deleteMany({ _room: roomId })
  }
}
